### PR TITLE
[air] `pyarrow.fs` persistence (13/n): Support the `Result.from_path` API

### LINKED ISSUE #38579: [train][GA] Support `Result.from_path` w/o `.tune_metadata` file
* Add a new auto-filled metric that is the current checkpoint index.
* Should be None or -1 if no checkpoint was reported at that step.
* This is needed to be able to get the metrics corresponding to each checkpoint from the result.json file.
* Otherwise, you need the trial object, which is at the experiment dir level.

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

We no longer save `.tune_metadata` (which contained the metrics associated with a checkpoint) in the checkpoint directory. `Result.from_path` used to depend on this file to pull the metrics for each checkpoint (to populate `Result.best_checkpoints`). This PR adds a new auto-filled metric checkpoint `checkpoint_dir_name` that is equal to the persisted checkpoint directory name (e.g., `checkpoint_00003`) if a checkpoint was saved. Otherwise, this metric is `None`. Then, `Result.from_path` is updated to read from the `result.json` and match the metrics corresponding to the checkpoint directory, instead of `.tune_metadata`.

## Related issue number

Closes https://github.com/ray-project/ray/issues/38579

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]
It seems the description is out of date, but listing the files sounds fine to me as well.

**Comment by justinvyu** [comment]
Which description is out of date?

**Comment by ericl** [comment]
Seems we don't read any checkpoint names from the json.

File: python/ray/air/result.py [hunk 1]
```diff
@@ -166,42 +170,46 @@ def from_path(cls, path: str) -> "Result":
...
+                checkpoints = [
+                    Checkpoint.from_directory(checkpoint_dir)
+                    for checkpoint_dir in checkpoint_dirs
+                ]
+
+            metrics = []
+            for checkpoint_dir in checkpoint_dirs:
+                metrics_corresponding_to_checkpoint = metrics_df[
+                    metrics_df[CHECKPOINT_DIR_NAME] == checkpoint_dir.name
+                ]
```
Comment (justinvyu) [inline python/ray/air/result.py:191]: Oh, the `metrics_df` got loaded from the json, and I'm just reading the checkpoint dir name column from it to match w/ the checkpoint dir name.

Comment (justinvyu) [reply python/ray/air/result.py:191]: @ericl

Comment (ericl) [reply python/ray/air/result.py:191]: Oh, got it!

**Review by matthewdeng — APPROVED** [review]

**Pull request merged**